src/benchmark/tpch_benchmark.cpp

```diff
@@ -0,0 +1,355 @@
+ * IndividualQueries and PermutedQuerySets. See docs on BenchmarkMode for details.
+ * The benchmark will stop issuing new queries if either enough iterations have taken place or enough time has passed.
+ *
+ * main() is mostly concerned with parsing the CLI options while TpchBenchmark::run() performs the actual benchmark
+ * logic.
+ */
+
+namespace opossum {
+
+/**
+ * IndividualQueries runs each query a number of times and then the next one
+ * PermutedQuerySets runs the queries as sets permuting their order after each run (this exercises caches)
+ */
+enum class BenchmarkMode { IndividualQueries, PermutedQuerySets };
+
+using Duration = std::chrono::high_resolution_clock::duration;
+using TimePoint = std::chrono::high_resolution_clock::time_point;
+
+// Used to config out()
+auto verbose_benchmark = false;
+
+/**
+ * @return std::cout if `verbose_benchmark` is true, otherwise returns a discarding stream
+ */
+std::ostream& out() {
+  if (verbose_benchmark) {
+    return std::cout;
+  }
+
+  // Create no-op stream that just swallows everything streamed into it
+  // See https://stackoverflow.com/a/11826666
+  class NullBuffer : public std::streambuf {
+   public:
+    int overflow(int c) { return c; }
+  };
+
+  static NullBuffer null_buffer;
+  static std::ostream null_stream(&null_buffer);
+
+  return null_stream;
+}
+
+struct QueryBenchmarkResult {
+  size_t num_iterations = 0;
+  Duration duration = Duration{};
+};
+
+using BenchmarkResults = std::unordered_map<size_t, QueryBenchmarkResult>;
+
+/**
+ * Loosely copying the functionality of benchmark::State
+ * keep_running() returns false once enough iterations or time has passed.
+ *
+ */
+struct BenchmarkState {
+  enum class State { NotStarted, Running, Over };
+
+  BenchmarkState(const size_t max_num_iterations, const Duration max_duration)
+      : max_num_iterations(max_num_iterations), max_duration(max_duration) {}
+
+  bool keep_running() {
+    switch (state) {
+      case State::NotStarted:
+        begin = std::chrono::high_resolution_clock::now();
+        state = State::Running;
+        break;
+      case State::Over:
+        return false;
+      default: {}
+    }
+
+    if (num_iterations >= max_num_iterations) {
+      end = std::chrono::high_resolution_clock::now();
+      state = State::Over;
+      return false;
+    }
+
+    end = std::chrono::high_resolution_clock::now();
+    const auto duration = end - begin;
+    if (duration >= max_duration) {
+      state = State::Over;
+      return false;
+    }
+
+    num_iterations++;
+
+    return true;
+  }
+
+  State state{State::NotStarted};
+  TimePoint begin = TimePoint{};
+  TimePoint end = TimePoint{};
+
+  size_t num_iterations = 0;
+  size_t max_num_iterations;
+  Duration max_duration;
+};
+
+// Stores config and state of a benchmark mode instantiation
+class TpchBenchmark final {
+ public:
+  TpchBenchmark(const BenchmarkMode benchmark_mode, const std::optional<std::string>& output_file_path,
```

Comment:
We usually have optional parameters at the end, no matter if they are C-style optional or STL optionals.

Reply:
ooookay... changed it!